internal/cmd/dockerbuild/main.go

```diff
@@ -0,0 +1,219 @@
+				pluginKey = owner + "/" + pluginToBuild.PluginVersion
+			}
+		case "protocolbuffers":
+			switch identity.Plugin() {
+			case "cpp", "csharp", "java", "kotlin", "objc", "php", "pyi", "python", "ruby":
+				pluginKey = owner + "/" + pluginToBuild.PluginVersion
+			}
+		default:
+			// Assume everything else can be built independently
+			pluginKey = identity.IdentityString()
+		}
+		pluginGroups[pluginKey] = append(pluginGroups[pluginKey], pluginToBuild)
+	}
+	limit := runtime.GOMAXPROCS(0)
+	if limit > maxLimit {
+		limit = maxLimit
+	}
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	eg.SetLimit(limit)
+	for _, plugins := range pluginGroups {
+		plugins := plugins
+		eg.Go(func() error {
+			for _, pluginToBuild := range plugins {
+				log.Println("building:", pluginToBuild.Name, pluginToBuild.PluginVersion)
+				start := time.Now()
+				output, err := dockerBuild(ctx, pluginToBuild, dockerOrg, args)
+				if err != nil {
+					if errors.Is(err, context.Canceled) || strings.Contains(err.Error(), "signal: killed") {
+						return err
+					}
+					return fmt.Errorf(
+						"failed to build %s:%s: %w\noutput:\n%s",
+						pluginToBuild.Name,
+						pluginToBuild.PluginVersion,
+						err,
+						string(output),
+					)
+				}
+				elapsed := time.Since(start)
+				log.Println("built:", pluginToBuild.Name, pluginToBuild.PluginVersion, "in", elapsed.Round(time.Second))
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func dockerBuild(ctx context.Context, plugin *plugin.Plugin, dockerOrg string, args []string) ([]byte, error) {
+	dockerCmd, err := exec.LookPath("docker")
+	if err != nil {
+		return nil, err
+	}
+	identity, err := bufpluginref.PluginIdentityForString(plugin.Name)
+	if err != nil {
+		return nil, err
+	}
+	commonArgs := []string{
+		"buildx",
+		"build",
+		"--label",
+		fmt.Sprintf("build.buf.plugins.config.owner=%s", identity.Owner()),
+		"--label",
+		fmt.Sprintf("build.buf.plugins.config.name=%s", identity.Plugin()),
+		"--label",
+		fmt.Sprintf("build.buf.plugins.config.version=%s", plugin.PluginVersion),
+		"--label",
+		"org.opencontainers.image.source=https://github.com/bufbuild/plugins",
+		"--label",
+		fmt.Sprintf("org.opencontainers.image.description=%s", plugin.Description),
+		"--label",
+		fmt.Sprintf("org.opencontainers.image.licenses=%s", plugin.SPDXLicenseID),
+		"--progress",
+		"plain",
+	}
+	commonArgs = append(commonArgs, args...)
+	buildStages, err := getDockerBuildStages(filepath.Join(filepath.Dir(plugin.Path), "Dockerfile"))
+	if err != nil {
+		return nil, err
+	}
+	for _, stage := range buildStages {
+		// Build each stage of multi-stage build (to improve caching)
+		cmd := exec.CommandContext(
+			ctx,
+			dockerCmd,
+			commonArgs...,
+		)
+		cmd.Args = append(
+			cmd.Args,
+			"--target",
+			stage,
+			"-t",
+			fmt.Sprintf("%s/plugins-%s-%s:%s-%s", dockerOrg, identity.Owner(), identity.Plugin(), plugin.PluginVersion, stage),
```

Comment:
By saving results of each build stage, we can ensure when building multiple versions of plugins like protoc/grpc ones, we'll save the build result from prior builds. Otherwise, each version build would stomp on each other.